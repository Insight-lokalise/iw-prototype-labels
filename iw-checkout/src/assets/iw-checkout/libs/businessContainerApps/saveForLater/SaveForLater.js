import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { ShowIf } from '../../higherOrderComponents'
import { t } from '@insight/toolkit-utils/lib/labels'
import { IWAnchor, IWButton } from '../../iw-components'
import SaveLineLevels, { linkTypes } from '../cart/components/lineLevelForm/saveLineLevels'
import {
    selector_hasUsagePeriodReportableNonReportable,
    selector_cart,
    selector_isCloudCart, selector_cartItemsEnrollment,
} from './../../Cart/selectors'
import { selector_isLimitedUser } from './../../User/selectors'
import { userPermissions } from './../../User/permissions'
import { selector_hasUserPermission } from './../../User/selectors'
import SaveForLaterModal from './SaveForLaterModal'
import {submitEnrollmentValues} from "../cart/actions";
import {enrollmentInfoToUpdate} from './../../businessContainerApps/cart/helpers'

export class SaveForLater extends Component {
    constructor() {
        super()
        this.state = {
            saveCartModalIsOpen: false,
            orderTemplateModalIsOpen: false,
        }
        this.openSaveCartModal = this.openModal.bind(this, 'saveCart')
        this.openOrderTemplateModal = this.openModal.bind(this, 'orderTemplate')
    }


    closeModal = () => {
        this.setState({ saveCartModalIsOpen: false, orderTemplateModalIsOpen: false })
    };

    openModal(type) {
        const { hasEnrollmentInfoToUpdate, optInPartners, optOutPartners} = enrollmentInfoToUpdate(this.props.enrollmentInfo)
        hasEnrollmentInfoToUpdate &&  this.props.submitEnrollmentValues({enrollments: optInPartners , removedIds: optOutPartners })
        if (type === 'saveCart') {
            this.setState({ saveCartModalIsOpen: true, orderTemplateModalIsOpen: false })
        } else if (type === 'orderTemplate') {
            this.setState({ saveCartModalIsOpen: false, orderTemplateModalIsOpen: true })
        }
    }

    saveQuote = () => {
        const { hasEnrollmentInfoToUpdate, optInPartners, optOutPartners} = enrollmentInfoToUpdate(this.props.enrollmentInfo)
        hasEnrollmentInfoToUpdate ?
            this.props.submitEnrollmentValues({enrollments: optInPartners , removedIds: optOutPartners }).then(() => window.location =  "/insightweb/displaysavequote" )
            : window.location =  "/insightweb/displaysavequote"
    };

    render() {
        const saveForLaterText = t('Save for later')
        const saveAsQuoteText = t('Save as quote')
        const saveCartContentsText = t('Save cart contents')
        const saveOrderTemplateText = t('Save order template')
        const saveCart = 'save_cart'
        const viewQuotes = 'view_quotes'
        const {
            cartIsEmpty,
            canShowSaveForLaterSection,
            enrollmentInfo,
            isCloudCart,
            isLimitedUser,
            hasUsagePeriodReportableNonReportable,
            saveLineLevels,
            submitEnrollmentValues,
        } = this.props
        const {
            saveCartModalIsOpen,
            orderTemplateModalIsOpen,
        } = this.state
        return (
            !cartIsEmpty && canShowSaveForLaterSection && !isCloudCart &&
                <div className="save-for-later">
                    <div className="row section__header align-middle is-collapse-child hide-for-print">
                        <h3 className="columns section__header-title">
                            {saveForLaterText}
                        </h3>
                    </div>
                    <div className="row is-collapse-child hide-for-print">
                        <div className="columns">
                            { !isLimitedUser &&
                                 <ShowIf permissions={viewQuotes} test={!hasUsagePeriodReportableNonReportable}>
                                    { saveLineLevels ?
                                            <SaveLineLevels linkType={linkTypes.SAVE_AS_QUOTE_LINK} text={saveAsQuoteText} />
                                        :
                                            <IWAnchor className="save-for-later__action" onClick={this.saveQuote}  role="button" tabindex="0">
                                                {saveAsQuoteText}
                                            </IWAnchor>
                                    }
                                </ShowIf>
                            }
                            <ShowIf permissions={saveCart}>
                                <span>
                                    <IWButton className="save-for-later__action iw-button--link" onClick={this.openSaveCartModal}>
                                        {saveCartContentsText}
                                    </IWButton>
                                    <IWButton className="save-for-later__action iw-button--link" onClick={this.openOrderTemplateModal}>
                                        {saveOrderTemplateText}
                                    </IWButton>
                                    <SaveForLaterModal
                                        closeModal={this.closeModal}
                                        enrollmentInfo={enrollmentInfo}
                                        saveCartModalIsOpen={saveCartModalIsOpen}
                                        submitEnrollmentValues={submitEnrollmentValues}
                                        orderTemplateModalIsOpen={orderTemplateModalIsOpen}
                                    />
                                </span>
                            </ShowIf>
                        </div>
                    </div>
                </div>
        )
    }
}


SaveForLater.propTypes = {
    saveLineLevels: PropTypes.bool.isRequired,
}


SaveForLater.defaultProps = {
    saveLineLevels: false,
}

const mapStateToProps = state => ({
    canShowSaveForLaterSection: selector_hasUserPermission(state, userPermissions.SAVE_CART) ||
        selector_hasUserPermission(state, userPermissions.VIEW_QUOTES),
    cartIsEmpty: selector_cart(state).totalCount === 0,
    enrollmentInfo: selector_cartItemsEnrollment(state),
    isCloudCart: selector_isCloudCart(state),
    isLimitedUser: selector_isLimitedUser(state),
    hasUsagePeriodReportableNonReportable: selector_hasUsagePeriodReportableNonReportable(state),
})

const mapDispatchToProps = dispatch => bindActionCreators({
    submitEnrollmentValues,
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(SaveForLater)
