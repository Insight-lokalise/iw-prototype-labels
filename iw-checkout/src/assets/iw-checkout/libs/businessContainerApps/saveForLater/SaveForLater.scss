.save-for-later {
    background: white;
    padding-bottom: 15px;

    @include breakpoint(small only) {
        margin-bottom: 10px;
    }

    &__title {
        background-color: $brand-mediumgray;
        color: white;
        padding-top: 5px;
        padding-bottom: 5px;
    }

    &__action {
        padding-top: 10px;
        float: left;
        clear: left;
        text-align: left;
    }

    &__backdrop + .iw-dialog {
        header div:nth-of-type(2) {
            position: relative;

            span {
                display: block;
                position: absolute;
                right: 20px;
                top: -16px;
            }
        }
    }


    &__modal-content {
        padding: 45px 15px 0px;
        
        .form__field {
            margin-top: 5px;
        }

        ~ footer {
            margin-bottom: 20px;

            div {
                justify-content: flex-start;
            }

            button {
                width: 145px!important;
            }
        }
    }

    &__modal-messages {
        padding: 45px 15px 0px;
        text-align: center;
    }

    &__modal-success {
        color: $brand-cyan
    }

    &__modal-checkbox {
        margin: 25px 0;
        font-weight: 100!important;

        input {
            /* important to overwrite an !important that is being used in the commoncss */
            margin-right: 10px!important;
        }
    }

    .section__header {
        border-bottom: none;
    }
}

.hide-for-medium .save-for-later {
    border: 1px solid $color-warm-gray-5;
}
