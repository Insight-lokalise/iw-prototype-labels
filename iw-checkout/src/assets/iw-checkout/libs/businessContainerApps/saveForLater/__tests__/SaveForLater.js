import React from 'react'
import { shallow } from 'enzyme'

import { SaveForLater, mapStateToProps } from './../SaveForLater'
import { jonKrone_lenovo } from './../../../../__fixtures__/jonKrone_lenovo'

jest.mock('@insight/)

describe('<SaveForLater />', () => {
    it('should match previous snapshot.', () => {
        const wrapper = shallow(
            <SaveForLater {...mapStateToProps(jonKrone_lenovo)} />
        )
        expect(wrapper).toMatchSnapshot()
    })
})
