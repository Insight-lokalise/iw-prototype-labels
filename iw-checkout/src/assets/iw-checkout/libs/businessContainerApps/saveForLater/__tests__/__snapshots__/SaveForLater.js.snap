// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SaveForLater /> should match previous snapshot. 1`] = `
<Connect(ShowIfComponent)
  permissions="save_cart"
>
  <div
    className="save-for-later"
  >
    <div
      className="row is-collapse-child hide-for-print"
    >
      <h4
        className="columns save-for-later__title no-margin-bot"
      >
        Save for later
      </h4>
    </div>
    <div
      className="row is-collapse-child hide-for-print"
    >
      <div
        className="columns"
      >
        <Connect(ShowIfComponent)
          permissions="view_quotes"
          test={true}
        >
          <IWAnchor
            className="save-for-later__action"
            href="/insightweb/displaysavequote"
          >
            Save as quote
          </IWAnchor>
        </Connect(ShowIfComponent)>
        <Connect(ShowIfComponent)
          permissions="save_cart"
        >
          <span>
            <IWAnchor
              className="save-for-later__action"
              href="/insightweb/manageCart/Cart"
            >
              Save cart contents
            </IWAnchor>
            <IWAnchor
              className="save-for-later__action"
              href="/insightweb/manageCart/Template"
            >
              Save order template
            </IWAnchor>
          </span>
        </Connect(ShowIfComponent)>
      </div>
    </div>
  </div>
</Connect(ShowIfComponent)>
`;
